src/unlock/rc_data.rs

```diff
@@ -0,0 +1,230 @@
+    let proof_pairs: Vec<(SmtH256, SmtH256)> = proof_keys.iter().map(map_noexist).collect();
+
+    let smt = new_smt(pairs);
+    let root = smt.root();
+
+    let proof = smt
+        .merkle_proof(proof_pairs.clone().into_iter().map(|(k, _)| k).collect())
+        .map_err(|err| RcDataError::BuildTree(err.to_string()))?;
+    let compiled_proof = proof
+        .compile(proof_pairs)
+        .map_err(|e| RcDataError::CompileProof(e.to_string()))?;
+
+    Ok((*root, compiled_proof.into()))
+}
+
+fn build_rc_rule(smt_root: &[u8; 32], is_black: bool, is_emergency: bool) -> Bytes {
+    let mut flags: u8 = 0;
+
+    if !is_black {
+        flags ^= WHITE_BLACK_LIST_MASK;
+    }
+    if is_emergency {
+        flags ^= EMERGENCY_HALT_MODE_MASK;
+    }
+    let rcrule = RCRuleBuilder::default()
+        .flags(flags.into())
+        .smt_root(smt_root.pack())
+        .build();
+    let res = RCDataBuilder::default()
+        .set(RCDataUnion::RCRule(rcrule))
+        .build();
+    res.as_bytes()
+}
+
+#[repr(u8)]
+pub enum Mask {
+    Neither = 0,
+    Input = 1,
+    Output = 2,
+    Both = 3,
+}
+pub struct ProofWithMask {
+    pub proof: Vec<u8>,
+    pub mask: Mask,
+}
+
+impl ProofWithMask {
+    pub fn new(proof: Vec<u8>, mask: Mask) -> Self {
+        ProofWithMask { proof, mask }
+    }
+}
+
+pub fn build_proofs(proofs: Vec<ProofWithMask>) -> SmtProofEntryVec {
+    let mut builder = SmtProofEntryVecBuilder::default();
+    for ProofWithMask { proof, mask } in proofs {
+        let proof_builder = SmtProofBuilder::default().set(
+            proof
+                .iter()
+                .map(|v| molecule::prelude::Byte::new(*v))
+                .collect(),
+        );
+
+        let temp = SmtProofEntryBuilder::default()
+            .proof(proof_builder.build())
+            .mask((mask as u8).into());
+        builder = builder.push(temp.build());
+    }
+    builder.build()
+}
+
+pub fn generate_single_proof(
+    on: bool,
+    smt_key: &[SmtH256],
+    whitelist: bool,
+) -> Result<(Vec<u8>, Bytes)> {
+    let hash = smt_key;
+    let (smt_root, proof) = if whitelist {
+        build_smt_on_wl(hash, on)?
+    } else {
+        build_smt_on_bl(hash, on)?
+    };
+
+    let rc_rule = build_rc_rule(&smt_root.into(), !whitelist, false);
+    Ok((proof, rc_rule))
+}
+
+pub type RcProofWithRule = (Vec<ProofWithMask>, Vec<Bytes>);
+pub fn generate_proofs(smt_key: &[SmtH256], whitelist: bool) -> Result<RcProofWithRule> {
+    let mut proofs = Vec::<ProofWithMask>::default();
+    let mut rc_rules = Vec::<Bytes>::default();
+
+    let (proof1, rc_rule1) = generate_single_proof(true, smt_key, whitelist)?;
+    proofs.push(ProofWithMask::new(proof1, Mask::Input));
+    rc_rules.push(rc_rule1);
+
+    let (proof2, rc_rule2) = generate_single_proof(false, smt_key, whitelist)?;
+    proofs.push(ProofWithMask::new(proof2, Mask::Input));
+    rc_rules.push(rc_rule2);
+
+    Ok((proofs, rc_rules))
+}
```

Comment:
Please add Rust doc for public function/struct and add some unit tests for:
* `build_smt_on_bl()`
* `build_smt_on_wl()`
* `generate_single_proof()`
* `generate_proofs()`